core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -36,373 +73,288 @@ class Queue[A] private (
-                      IO.now(true) -> Surplus(
-                        IQueue.empty[A].enqueue(addToQueue.toList),
-                        IQueue.empty.enqueue(surplusValues -> p),
-                        hook
-                      )
-                    case Sliding =>
-                      val notTaken = addToQueue ++ surplusValues
-                      val toQueue  = capacity.fold(notTaken)(notTaken.takeRight)
-                      p.complete(false) -> Surplus(
-                        IQueue.empty.enqueue(toQueue.toList),
-                        IQueue.empty,
-                        hook
-                      )
-                    case Dropping =>
-                      p.complete(false) -> Surplus(
-                        IQueue.empty.enqueue(addToQueue.toList),
-                        IQueue.empty,
-                        hook
-                      )
-                  }
-
-              completeTakers *> complete -> surplus
-            } else
-              completeTakers *> p.complete(true) -> Deficit(deficitValues, hook)
-        }
-
-      case (p, Surplus(values, putters, hook)) =>
-        val (addToQueue, surplusValues) =
-          capacity.fold((as, Iterable.empty[A]))(c => as.splitAt(c - values.size))
-        if (surplusValues.isEmpty)
-          (p.complete(true), Surplus(values.enqueue(addToQueue.toList), putters, hook))
-        else {
-          strategy match {
-            case BackPressure =>
-              (
-                IO.now(true),
-                Surplus(values.enqueue(addToQueue.toList), putters.enqueue(surplusValues -> p), hook)
-              )
-            case Sliding =>
-              (
-                p.complete(true),
-                capacity.fold(
-                  Surplus(values.enqueue(as.toList), putters, hook)
-                ) { c =>
-                  Surplus(
-                    values
-                      .takeRight(c - as.size)
-                      .enqueue(as.takeRight(c).toList),
-                    putters,
-                    hook
-                  )
-                }
-              )
-            case Dropping =>
-              (
-                p.complete(false),
-                Surplus(values.enqueue(addToQueue.toList), putters, hook)
-              )
-          }
-        }
-
-      case (p, state @ Shutdown) => (p.interrupt, state)
-    }
-
-    val release: (Boolean, Promise[Nothing, Boolean]) => IO[Nothing, Unit] = {
-      case (_, p) => p.poll.void <> removePutter(p)
-    }
-
-    Promise.bracket[Nothing, State[A], Boolean, Boolean](ref)(acquire)(release)
-  }
+  final def offerAll(as: Iterable[A]): IO[Nothing, Boolean] =
+    for {
+      _                      <- checkShutdownState
+      takers                 <- IO.sync(unsafePollN(takers, as.size))
+      (forTakers, remaining) = as.splitAt(takers.size)
+      _                      <- IO.parTraverse(takers zip forTakers) { case (taker, item) => taker.complete(item) }
```

Comment:
I suggest using `traverse` instead of `parTraverse`, as it will be faster (each action only completes the promise, which is nearly instantaneous).